server/src/fstar_connection.ts

```diff
@@ -0,0 +1,512 @@
+
+	private handleResponse(msg: object) {
+		// Events for specific message types
+		if (isProtocolInfo(msg)) {
+			this.handleProtocolInfo(msg as ProtocolInfo);
+		} else {
+			// Either we expect unprompted protocol-info messages, or we expect
+			// responses to a query we sent.
+			const r = msg as IdeQueryResponse;
+			// TODO(klinvill): it appears that responses to full-buffer queries
+			// are sent with query ids with incrementing fractional components.
+			// E.g. if the query id is 2, the responses seem to come back with
+			// query ids 2, 2.1, 2.2, 2.3, etc. To deal with this behavior I
+			// truncate the responses query-id. Is this the proper behavior for
+			// this case?
+			const qid = Math.trunc(Number(r["query-id"]));
+			if (!qid) {
+				console.warn(`Ill-formed query response message: ${r}`);
+				return;
+			}
+
+			if (r.kind === 'message' && r.level === 'progress') {
+				const response = r.contents as IdeProgress;
+				// Progress responses are sent as a stream of messages. The last
+				// message is always "full-buffer-finished".
+				const done = response.stage === 'full-buffer-finished';
+				this.respondStream(qid, response, done);
+			} else if (r.kind === 'message' && r.level === 'proof-state') {
+				// TODO(klinvill): What queries can prompt proof-state status messages? Are they part of a stream (like responses to full-buffer queries)?
+				this.respond(qid, r.contents as IdeProofState);
+			} else if (r.kind === 'message' && r.level === 'info') {
+				// TODO(klinvill): info messages are just logged and don't resolve the corresponding pending_responses promise. Is this the right behavior (that info messages are just extraneous and should be logged and ignored)?
+				console.log("Info: " + r.contents);
+			} else if (r.kind === "response" && r.response) {
+				const responseType = FStarConnection.decideIdeReponseType(r.response);
+				if (responseType === 'symbol') {
+					const pr = this.pending_responses.get(qid);
+					if (!pr) {
+						console.warn(`No inflight query found for query-id: ${qid}, got response: ${JSON.stringify(r.response)}.`);
+						return;
+					}
+					// TODO(klinvill): can symbol messages be sent in response
+					// to a request that results in streams (like full-buffer
+					// queries)? I seem to be seeing this for some full-buffer
+					// queries. I therefore check below if the response is part
+					// of a stream and handle it accordingly.
+					if (pr.is_stream) {
+						// TODO(klinvill): Can a symbol message end a full-buffer stream? Or will it always only end with a full-buffer-finished message?
+						this.respondStream(qid, r.response, false);
+					}
+					else {
+						this.respond(qid, r.response);
+					}
+				} else if (responseType === 'error') {
+					// TODO(klinvill): Should we instead reject the promise
+					// here? Or pass along the error responses as the resolved
+					// promise value?
```

Comment:
This is a good question.  For example, the lookup query returns an error if it cannot resolve the identifier.  That is not really something that I would use an exception for.  So I'd tend to have a `Promise<IdeQueryResponse | IdeError>` type.

Reply:
I update the return types for the fullBufferQuery requests to have the type `Promise<IdeProgress | IdeErrors>`. Are there other methods that could result in IdeError responses being sent? On the protocol wiki page (https://github.com/FStarLang/FStar/wiki/Editor-support-for-F*#list-of-status-messages) I only see this being the case for `push` requests.